spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java

```diff
@@ -0,0 +1,188 @@
+
+/**
+ * A basic implementation of {@link RetryOperations} that uses a
+ * {@link RetryPolicy} and a {@link BackOffPolicy} to retry a
+ * {@link RetryCallback}. By default, the callback will be called
+ * 3 times with a fixed backoff of 1 second.
+ *
+ * <p>It is also possible to register a {@link RetryListener} to intercept and inject code
+ * during key retry phases (before a retry attempt, after a retry attempt, etc.).
+ *
+ * <p>All retry operations performed by this class are logged at debug level,
+ * using "org.springframework.core.retry.RetryTemplate" as log category.
+ *
+ * @author Mahmoud Ben Hassine
+ * @since 7.0
+ * @see RetryOperations
+ * @see RetryPolicy
+ * @see BackOffPolicy
+ * @see RetryListener
+ */
+public class RetryTemplate implements RetryOperations {
+
+	protected final Log logger = LogFactory.getLog(getClass());
+
+	private RetryPolicy retryPolicy = new MaxRetryAttemptsPolicy();
+
+	private BackOffPolicy backOffPolicy = new FixedBackOffPolicy(1000);
+
+	private RetryListener retryListener = new RetryListener() {
+	};
+
+	/**
+	 * Create a new retry template with default settings.
+	 */
+	public RetryTemplate() {
+	}
+
+	/**
+	 * Create a new retry template with a custom {@link RetryPolicy}.
+	 * @param retryPolicy the retry policy to use
+	 */
+	public RetryTemplate(RetryPolicy retryPolicy) {
+		Assert.notNull(retryPolicy, "Retry policy must not be null");
+		this.retryPolicy = retryPolicy;
+	}
+
+	/**
+	 * Create a new retry template with a custom {@link RetryPolicy} and {@link BackOffPolicy}.
+	 * @param retryPolicy the retry policy to use
+	 * @param backOffPolicy the backoff policy to use
+	 */
+	public RetryTemplate(RetryPolicy retryPolicy, BackOffPolicy backOffPolicy) {
+		this(retryPolicy);
+		Assert.notNull(backOffPolicy, "BackOff policy must not be null");
+		this.backOffPolicy = backOffPolicy;
+	}
+
+	/**
+	 * Set the {@link RetryPolicy} to use. Defaults to <code>MaxAttemptsRetryPolicy()</code>.
+	 * @param retryPolicy the retry policy to use. Must not be <code>null</code>.
+	 */
+	public void setRetryPolicy(RetryPolicy retryPolicy) {
+		Assert.notNull(retryPolicy, "Retry policy must not be null");
+		this.retryPolicy = retryPolicy;
+	}
+
+	/**
+	 * Set the {@link BackOffPolicy} to use. Defaults to <code>FixedBackOffPolicy(Duration.ofSeconds(1))</code>.
+	 * @param backOffPolicy the backoff policy to use. Must not be <code>null</code>.
+	 */
+	public void setBackOffPolicy(BackOffPolicy backOffPolicy) {
+		Assert.notNull(backOffPolicy, "BackOff policy must not be null");
+		this.backOffPolicy = backOffPolicy;
+	}
+
+	/**
+	 * Set the {@link RetryListener} to use. Defaults to a <code>NoOp</code> implementation.
+	 * If multiple listeners are needed, use a {@link CompositeRetryListener}.
+	 * @param retryListener the retry listener to use. Must not be <code>null</code>.
+	 */
+	public void setRetryListener(RetryListener retryListener) {
+		Assert.notNull(retryListener, "Retry listener must not be null");
```

Comment:
@sbrannen , @jhoeller , do we still do these asserts if we promote nullness?

Reply:
Sometimes, but not always. 

We can discuss the pros and cons offline.

Reply:
I kept the check for now. I will defer the decision to the team about conventions here.